pkg/pub_worker/test/analyze_test.dart

```diff
@@ -43,14 +46,114 @@ void main() {
+            'version': '$version',
+            'dependencies': dependencies,
+            'environment': {'sdk': '>=2.12.0-0 <3.0.0'},
+          },
+        };
+
     final app = Router();
 
+    app.get('/api/packages/retry', (Request request) {
+      return Response.ok(json.encode({
+        'name': 'retry',
+        'latest': packageVersion('retry', '3.0.1', {'meta': '^1.0.0'}),
+        'versions': [
+          packageVersion('retry', '3.0.1', {'meta': '^1.0.0'}),
+          packageVersion('retry', '3.0.0', {'meta': '^1.0.0'}),
+        ],
+      }));
+    });
+
+    app.get('/api/packages/meta', (Request request) {
+      return Response.ok(json.encode({
+        'name': 'meta',
+        'latest': packageVersion('meta', '1.0.0', {}),
+        'versions': [packageVersion('meta', '1.0.0', {})],
+      }));
+    });
+
+    app.get('/packages/retry/versions/3.0.0.tar.gz', (request) {
+      return Response.ok(Stream<TarEntry>.fromIterable([
+        TarEntry.data(
+          TarHeader(name: 'pubspec.yaml'),
+          json.fuse(utf8).encode({
+            'name': 'retry',
+            'version': '3.0.0',
+            'dependencies': {
+              'meta': '^1.0.0',
+            },
+            'environment': {'sdk': '>=2.12.0-0 <3.0.0'},
+          }),
+        ),
+        TarEntry.data(
+          TarHeader(name: 'lib/retry.dart'),
+          utf8.encode('''
+            void retry(void Function() fn) { for (var i = 0; i < 5; i++) fn();}
+          '''),
+        ),
+      ]).transform(tarWriter).transform(gzip.encoder));
+    });
+
+    app.get('/packages/retry/versions/3.0.1.tar.gz', (request) {
+      return Response.ok(Stream<TarEntry>.fromIterable([
+        TarEntry.data(
+          TarHeader(name: 'pubspec.yaml'),
+          json.fuse(utf8).encode({
+            'name': 'retry',
+            'version': '3.0.1',
+            'dependencies': {
+              'meta': '^1.0.0',
+            },
+            'environment': {'sdk': '>=2.12.0-0 <3.0.0'},
+          }),
+        ),
+        TarEntry.data(
+          TarHeader(name: 'lib/retry.dart'),
+          utf8.encode('''
+            import 'package:meta/meta.dart';
+            @Sealed()
+            void retry(void Function() fn) { for (var i = 0; i < 5; i++) fn();}
+          '''),
+        ),
+      ]).transform(tarWriter).transform(gzip.encoder));
+    });
+
+    app.get('/packages/meta/versions/1.0.0.tar.gz', (request) {
+      return Response.ok(Stream<TarEntry>.fromIterable([
+        TarEntry.data(
+          TarHeader(name: 'pubspec.yaml'),
+          json.fuse(utf8).encode({
+            'name': 'meta',
+            'version': '1.0.0',
+            'dependencies': {},
+            'environment': {'sdk': '>=2.12.0-0 <3.0.0'},
+          }),
+        ),
+        TarEntry.data(
+          TarHeader(name: 'lib/meta.dart'),
+          utf8.encode('''
+            class Sealed {}
+          '''),
+        ),
+      ]).transform(tarWriter).transform(gzip.encoder));
+    });
```

Comment:
This is a moderate amount of madness, but it's really nice to have self contained tests like this :D

Reply:
Yeah, that's what I did before, but this wasn't so much work  --- and now it's nice and self-contained, no changes on pub.dev is going to break this test :D